wordpress-mobile/WordPress-iOS
# [Tooling] Re-implement AppLocalizedStrings linter in swift

This work is part of paaHJt-2Ib-p2, and a follow up of the annoucement in paNNhX-nP-p2.

## What

This re-implements the `AppLocalizedString` linter introduced in https://github.com/wordpress-mobile/WordPress-iOS/pull/18427 (which was written in Ruby) to re-write it in pure Swift

## Why

 - Because it's always nice to have one less tooling depending on Ruby
 - Because one of our mid-to-long term plan is to not require developers to install Ruby and `rbenv` on their Mac anyway (and even if we're not there yet, better not continue increasing our ruby usage for devs if possible)
 - Because I couldn't resist being nerd-snipe'd by @mokagio in https://github.com/wordpress-mobile/WordPress-iOS/pull/18427#pullrequestreview-954314088

## How

 - Since `project.pbxproj` files are in the ASCII-Plist format, they can easily be parsed and decoded using `PropertyListDecoder`
 - I thus implemented a couple of `Decodable` types to interpret the various structures present in a `.pbxproj` file (only the ones relevant to our needs)
 - The _main_ part of the swift script then use those `Decodable` models to parse the `.xcodeproj` file, iterate on its `PBXNativeTarget` objects of type `.appExtension`, find their build phases, resolve the path of each `PBXBuildFile` listed as child of the "Compile Sources" phase, and lint each of them in turn.
 - The Build Phase calling this linter script (~in each of the app extension target of the Xcode project~) calls a `LintAppLocalizedStringsUsage.sh` wrapper script which takes care of:
   - Pre-compiling the swift script into a binary (using the proper macOS SDK), to speed up further executions (instead of having to re-interpret the swift script each time)
   - Run the pre-compiled script on the appextension targets

## To Test

 - Build the project, validate it doesn't get any error while trying to run the lint Script Build Phase nor does trigger any violation
 - Optional: Go to the Build log, search for `[Lint]` to find all the instances invoking the linter script (there should be one invocation per app extension target), and verify that while the first invocation might have taken a couple of seconds, all the successive invocations used the pre-compiled version of the linter and took <1s
 - Introduce a violation on purpose, by replacing any call of `AppLocalizedString` in the codebase (in a file used by an app extension) by `NSLocalizedString`
 - Build the project again, and check that the violation was flagged as an error in your IDE
 - fix the violation you introduced on purpose, then build the project again, and check the error disappeared

#### Optional:
 - Go to the Build log, search for `[Lint]` and verify that all the invocations of the linter were fast and used the pre-compiled version of the linter.
 - Edit the `Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift` file to add a space somewhere (so it is seen as changed)
 - Build the project again, and verify that the first invocation of the script in a `[Lint]` phase detected the source of the linter changed and re-compiled it from the new source before proceeding
